cmd/juju/controller/addmodel.go

```diff
@@ -405,49 +417,82 @@ func (c *addModelCommand) maybeUploadCredential(
 	modelOwner string,
 ) (names.CloudCredentialTag, error) {
 
-	modelOwnerTag := names.NewUserTag(modelOwner)
-	credentialTag, err := common.ResolveCloudCredentialTag(
-		modelOwnerTag, cloudTag, c.CredentialName,
-	)
-	if err != nil {
-		return names.CloudCredentialTag{}, errors.Trace(err)
+	// If the user has not specified a credential, and the cloud advertises
+	// itself as supporting the "empty" auth-type, then return immediately.
+	if c.CredentialName == "" {
+		for _, authType := range cloud.AuthTypes {
+			if authType == jujucloud.EmptyAuthType {
+				return names.CloudCredentialTag{}, nil
+			}
+		}
 	}
 
 	// Check if the credential is already in the controller.
-	//
-	// TODO(axw) consider implementing a call that can check
-	// that the credential exists without fetching all of the
-	// names.
+	modelOwnerTag := names.NewUserTag(modelOwner)
 	credentialTags, err := cloudClient.UserCredentials(modelOwnerTag, cloudTag)
 	if err != nil {
 		return names.CloudCredentialTag{}, errors.Trace(err)
 	}
-	credentialId := credentialTag.Id()
-	for _, tag := range credentialTags {
-		if tag.Id() != credentialId {
-			continue
+	if c.CredentialName != "" {
+		credentialTag, err := common.ResolveCloudCredentialTag(
+			modelOwnerTag, cloudTag, c.CredentialName,
+		)
+		if err != nil {
+			return names.CloudCredentialTag{}, errors.Trace(err)
+		}
+		credentialId := credentialTag.Id()
+		for _, tag := range credentialTags {
+			if tag.Id() != credentialId {
+				continue
+			}
+			ctx.Infof("Using credential '%s' cached in controller", c.CredentialName)
+			return credentialTag, nil
+		}
+		if credentialTag.Owner().Id() != modelOwner {
+			// Another user's credential was specified, so
+			// we cannot automatically upload.
+			return names.CloudCredentialTag{}, errors.NotFoundf(
+				"credential '%s'", c.CredentialName,
+			)
 		}
-		ctx.Infof("Using credential '%s' cached in controller", c.CredentialName)
-		return credentialTag, nil
 	}
 
-	if credentialTag.Owner().Id() != modelOwner {
-		// Another user's credential was specified, so
-		// we cannot automatically upload.
-		return names.CloudCredentialTag{}, errors.NotFoundf(
-			"credential '%s'", c.CredentialName,
-		)
+	// The user has either not specified a credential, or if they have, it
```

Comment:
Yes it does:
 - `The user has either not specified a credential, or ...`
 - `if c.CredentialName == "" &&`